fix: npm should be supported only if package-lock.json exists

carlosthe19916 opened the pull request:
## Description

In the context of the npm provider:
- If a project uses Yarn, then the current provider forces `npm install` therefore creates necessarily a `package-lock.json` file  not respecting the package manager chosen by the user.
- This PR ensures the NPM provider is supported only if `package.json` exists AND `package-lock.json` exists. This way we respect the user package manager. Moreover, `packace-lock.json` is supposed to give us the exact set of packages used by a project. If a project does not have that file then the package scanning won't be idempotent neither reproducible. 

**Related issue (if any):** fixes https://github.com/fabric8-analytics/fabric8-analytics-vscode-extension/issues/738

## Checklist

- [x] I have followed this repository's contributing guidelines.
- [x] I will adhere to the project's code of conduct.

## Additional information

I think this problem should be addressed in two steps:
1. Make sure the current NPM provider relies on `package-lock.json`. This step is done in this PR
2. Create an additional provider for YARN which should rely on `yarn.lock`. Additionally we can go beyond and also support PNPM which relies on `pnpm-lock.yaml`. 

Just like in the Java World we required Maven and Gradle, in the Javascript World we will require NPM, Yarn and PNPM.

zvigrinberg reviewed (changes requested):
Hi @carlosthe19916 
The approach looks good to me in general, but i have two comments:

1. As we don't support yarn and do support npm, then we should verify that the yarn.lock file doesn't exists in manifest dir in order to proceed with npm.
We do support to perform both component and stack analysis , even if the user didn't install the application ( didn't run npm install, in such case, no package-lock.json file exists), thus , the assumption should be that it's npm , unless you have yarn.lock ( indicating that the user is using yarn as package manager instead of npm), and thus it's not supported.

2. This interesting approach to block yarn projects from being analysed  , introduced another problem
 https://github.com/RHEcosystemAppEng/exhort-javascript-api/blob/7c29e74da7829386b8c1d70416cb3b867559cdb7/src/provider.js#L33 
 in which , in case it is a project managed by yarn package manager, it will  print 
`package.json` is not supported, which is not true ( as it's supported with npm obviously), and it should be something like "`npm.json` is not supported for yarn package manager"

But since this provider type is sort of interface/abstract class in notion with a sort of template method design pattern( the `match` function), it should remain generic without insert in it specific logic that is related to npm/yarn.

you should think of a way how to address it, as right now, the error message is not correct, misleading, and not indicating the real reason why package.json is not supported ( not supported with yarn being the package manager).

carlosthe19916 commented:
@zvigrinberg thank you very much for your valuable review. 

> you should think of a way how to address it, as right now, the error message is not correct, misleading, and not indicating the real reason why package.json is not supported ( not supported with yarn being the package manager).

I agree. I will fix it.

> We do support to perform both component and stack analysis , even if the user didn't install the application ( didn't run npm install, in such case, no package-lock.json file exists), thus , the assumption should be that it's npm , unless you have yarn.lock ( indicating that the user is using yarn as package manager instead of npm), and thus it's not supported.

I gave a lot of thought to this while reviewing the issues reported at https://github.com/fabric8-analytics/fabric8-analytics-vscode-extension

People are complaining that their package-lock.json or package.son are being manipulated. I verified that if a project only has `package.json` and not a dependency lock file (of any type); then because we assume it is an NPM project we generate automatically the `package-lock.json` in their projects, which is technically manipulating their file system without their consent.

This is what I had originally in my mind: Only support `package.json` as long as their dependency lock file exists e.g.

| Package manager | manifest     | dependency lock file | action                 |
|-----------------|--------------|----------------------|------------------------|
| npm             | package.json | package-lock.json    | Supported in this PR   |
| yarn            | package.json | yarn.lock            | Cover it in another PR |
| pnpm            | package.json | pnpm-lock.yaml       | Cover it in another PR |
| bum             | package.json | bun.lockb            | Cover it in another PR (not sure about this though) |

And my original idea was to always require `package-lock.json|yarn.lock|pnpm-lock.yaml|bun.lockb` because those files allow us to explicitly know which dependencies the user is using. Without a lock file We could not tell deterministically which packages a project is using (we can guess forcing a `npm install` though). But I might be wrong.

I was trying to avoid a logic like `if yarn.lock exists || pnpm-lock.json exists || bun.lockb exits` then assume it is `npm`.

For compatibility reasons I think I can still assume "if only package.json exists" then assume it is npm but it will generated a `package-lock.json` file without the consent of the user.

What do you think about the points I mentioned? I am not trying to push back, just trying to validate, or invalidate, my original idea with a more experienced guy like you :)

zvigrinberg commented:
I can see you point...
But there are  heavyconsequences - if user wouldn't install their node project using npm install, then it'll not perform both dependency analytics report generation ( stack analysis) and component analysis in the vscode editor when the editor is opened on package.json, And user will don't know what happened - unless you're adding to the error message statement that says something like  "package-lock.json is a requirement for ivoking analysis for npm projects."
but then , there are github action and tekton task that are using this lib, and they're not installing the manifest before invoking the analysis... ( they're being used as a stage/step in CI/CD pipelines/workflows.
So if you want to change according to your plan, then you need to change in all places ( tekton task, github action, vscode extension, lsp vscode extension, etc).
other than that, you'll need to change documentation of all plugins and extensions that are using this, and also the documentation of this library....

I Think it's better, and much easier doing it according to my suggestion.
In addition to that change in the lib, you can add to the documentation a note/warning, that yarn users must have yarn.lock file in the manifest dir,  in order to prevent side effects (such as that  npm will manipulate their package.json and breaking the application state) from happening, and emphasize in the docs that yarn is not supported, so if using yarn and the extension is being installed for the sake of other projects with npm or other ecosystems ( `go` , java - `mvn` or `gradle` , `pyhon pip` ), then it's recommended to have yarn projects with yarn.lock in the project root dir.

carlosthe19916 commented:
@ruromero I added some changes to my PR. 

- "npm" analysis will be  supported if:
  - There is a package-lock.json file next to package.json OR
  -  There isn't any lock file. E.g. none of package-lock.json, yarn.lock, or pnpm-lock.yaml exists. This way this PR does not break any already existing documentation or pipelines. We can change it gradually though (in future PRs).

This PR should avoid Yarn projects to be analyzed and avoid creating the file package-lock.json for projects that use Yarn. And should cover https://github.com/fabric8-analytics/fabric8-analytics-vscode-extension/issues/738

There was a valid concern from Zvi about the accuracy of the error message `{manigest} is not supported` and I changed it to `{manifest} or lockfile is not supported` just to keep it generic. 

WDYT?

ruromero reviewed (changes requested):
Thanks for the changes. I have one suggestion to help users understand what's needed.

carlosthe19916 commented on src/provider.js:
```diff
@@ -31,5 +31,5 @@ export function match(manifest, providers) {
 		return provider
 	}
-	throw new Error(`${manifestPath.base} or lockfile is not supported`)
+	throw new Error(`${manifestPath.base} is not supported or lock file does not exists`)
```
Good idea. It has been done in my latest commit.

ruromero commented on src/provider.js:
```diff
@@ -28,8 +28,16 @@ export const availableProviders = [new Java_maven(), new Java_gradle_groovy(), n
 	}
-	throw new Error(`${manifestPath.base} is not supported`)
+
+	if (provider.requireLockFile()) {
```
here you can just require a `validateLockFile` where default implementations will just return `undefined` or `null` and other implementations can return a specific error message like:
> `package.json` requires a lock file. Use your package manager to initialize it. Example: npm install

carlosthe19916 replied:
@ruromero thanks, I added some commits. Now I am using `validateLockFile` but it does not return anything but it is a `void` function. And in case the lock file is not supported then a throw error i generated by the provider itself, this way the provider can customize the Error message according to the technology it handles. WDYT?

ruromero reviewed (changes requested):
Thanks Carlos. Just a small remark.

ruromero reviewed (approved):
LGTM
